### ISSUE TITLE
Naming strategy causes poor scaling of MH+recursive models

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#150]
With models which build structure recursively, (Hash)MH run time doesn't appear to scale linearly with the number of ERP (n) as I expected:

![current](https://cloud.githubusercontent.com/assets/9109012/8706727/447bbc52-2b29-11e5-928d-e5fdc1654d36.png)

I think there are a couple of reasons for this:

1. In recursive models stack addresses grow with model size. I think this makes `concat`ing the result of `gensym` onto the current address O(n), leading to O(n^2) per iteration.
2. I speculate that V8 uses a fancy data structure for strings which delays computing the result of a `concat` as long as possible. In our case I think this means every time we look at `address` in MH's `sample` we trigger the concatenation of lots of tiny fragments of address, many of which we've already concatenated elsewhere.

To test this idea I tweaked the naming scheme to hash each address with MD5 after the `gensym` is concatenated to the current address. This probably isn't suitable for all of our needs, but I think it side steps both problems I mentioned. Here's what happened:

![hash](https://cloud.githubusercontent.com/assets/9109012/8706739/4e70f86c-2b29-11e5-8455-83f7736cc968.png)

Here's the same plot with a different y limit:

![hash2](https://cloud.githubusercontent.com/assets/9109012/8706742/5344be3c-2b29-11e5-94a1-9e27f7a96fb0.png)

I also tried forcing the `concat` to happen immediately (but without the hashing) by replacing `address.concat(gensym())` with something like `[address, gensym()].join('')`. This helps, presumable because it addresses problem 2 (?) but the hashing addresses approach does better:

![hash_vs_force_join](https://cloud.githubusercontent.com/assets/9109012/8706750/5ae8359c-2b29-11e5-8643-c3c3a216ca90.png)

The forced join approach seg faulted at ~15K random variables.

I'm not sure whether this plot supports the O(n^2) idea I mention in problem 1 or not.

Small print:

* [Benchmark code](https://github.com/null-a/webppl/commit/8d1fa04176741c9e1dd3ed13a26f050018785376)
* [Address hashing code](https://github.com/null-a/webppl/commit/f7218ea33de0aa28126567f51ad6e8c1c639dea3)
* All the inference tests still pass.
* In all MH benchmarks I deterministically picked the middle ERP to regen from in order to reduce the noise in the timings. I included a version of HashMH which doesn't exit early when the same value is re-sampled (at the regen point) for the same reason.

**Comment by stuhlmueller** [comment]
Very interesting!

I'm wondering if we can improve upon hashing both with respect to speed and collision probability. What if we replace `md5Concat` with `hashConcat` that looks something like this:

    var intHashCount = 0;
    var intHashTable = {};
    
    var hashConcat = function(a, b){
      var hash = intHashTable[[a, b]];
      if (hash === undefined){
        hash = intHashCount;
        intHashTable[[a, b]] = intHashCount;
        intHashCount += 1;    
      }
      return hash;
    }

Would that work?

For both this approach and the md5 approach, is there any effect on models without many variables? Does the test suite runtime change?

**Comment by dritchie** [comment]
This is surprising to me. I was under the impression that V8 uses string 
interning (i.e. hash consing), so that every time it goes to concat two 
strings it's already concat'ed, that operation should be memoized (and 
thus constant time). It might be worthwhile to try a quick isolated 
experiment to see if repeatedly concat'ing the same strings noticeably 
slows down as (one of) the strings get larger.

-- D

**Comment by null-a** [comment]
> Does the test suite runtime change?

Yes, it's slower. IIRC it takes ~80 secs rather than ~50 secs.

**Comment by null-a** [comment]
Here's [a small program](http://jsfiddle.net/f8rmmgkh/1/) which I think captures one aspect of this behavior.

In node this appears to scale in a similar way to my original experiment. Interestingly io.js and Chrome do a lot better:

![plot](https://cloud.githubusercontent.com/assets/9109012/9091476/70933156-3b98-11e5-9f27-df326a74a355.png)

In io.js and Chrome it still appears that this scales like O(n^2), albeit with a fairly small constant. This doesn't seem strange to me since addresses also grow with n. So even if hash consing gives constant time concats hashing the string to use as a key for the object (or whatever similar thing V8 does) seems likely to take work which grows with n, which combined with the number of iterations gives n^2 overall. Am I missing something?

@stuhlmueller Would hashing addresses or similar work with isosmc? It looks like you're exploiting the structure in the current addressing scheme.

**Comment by dritchie** [comment]
Here's some relevant info:
http://stackoverflow.com/questions/5276915/do-common-javascript-implementations-use-string-interning

It looks like while V8 does intern string literals, it does not intern the results of string operations (e.g. concat). Under these assumptions, the n^2 performance make sense.

As for why the constant factor is so much bigger for node, I'm not sure. One thing that might be happening: the `d[address];` line in your program has no effect, so it's possible that the JIT can get rid of it. It may be that (for some reason) Chrome and io.js are doing that, but node.js isn't. Anyway, that's a pretty hand-wavy hypothesis, so I wouldn't put too much stock in it.

**Comment by null-a** [comment]
> Here's some relevant info:

Thanks!

> the d[address]; line in your program has no effect, so it's possible that the JIT can get rid of it

Good idea, but if you remove it all 3 run in 0-1ms for all n so I don't think they are optimizing it away.

**Comment by stuhlmueller** [comment]
> Would hashing addresses or similar work with isosmc? It looks like you're exploiting the structure in the current addressing scheme.

Yes, hashing would break the way it currently works. I wouldn't be too sad about this, since the use of addresses in the current implementation of isosmc feels hacky in the first place. That said, `hashConcat` above is reversible, and could be extended to be more easily reversible by also keeping an inverse hash table `intHashCount -> [a, b]` or something like that.

**Comment by null-a** [comment]
I re-ran the test from this [comment](https://github.com/probmods/webppl/issues/150#issuecomment-128068264) using Node 4.1.0:

![plot](https://cloud.githubusercontent.com/assets/9109012/10014987/298ce700-6115-11e5-8d81-38dc668782c1.png)

Presumably it's the V8 update that has led to Node coming into line with io.js and Chrome here.

**Comment by stuhlmueller** [comment]
What does this mean for the original test—is MH runtime sufficiently linear in the number of choices that we can close this issue and recommend that people use the latest node version?

**Comment by null-a** [comment]
> What does this mean for the original test—is MH runtime sufficiently linear in the number of choices that we can close this issue and recommend that people use the latest node version?

I was hoping so, but I was waiting until I could re-run the original test and not just the simplified test case. I've now done that, and unfortunately the problem persists with node v4.1.0.

Sorry my previous comment was vague and premature.
